internal/controller/etcdcluster_controller.go

```diff
@@ -36,24 +44,349 @@ type EtcdClusterReconciler struct {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+			Labels: map[string]string{
+				"app.kubernetes.io/name":       "etcd",
+				"app.kubernetes.io/instance":   cluster.Name,
+				"app.kubernetes.io/managed-by": "etcd-operator",
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{Name: "peer", TargetPort: intstr.FromInt32(2380), Port: 2380, Protocol: corev1.ProtocolTCP},
+				{Name: "client", TargetPort: intstr.FromInt32(2379), Port: 2379, Protocol: corev1.ProtocolTCP},
+			},
+			Type:      corev1.ServiceTypeClusterIP,
+			ClusterIP: "None",
+			Selector: map[string]string{
+				"app.kubernetes.io/name":       "etcd",
+				"app.kubernetes.io/instance":   cluster.Name,
+				"app.kubernetes.io/managed-by": "etcd-operator",
+			},
+			PublishNotReadyAddresses: true,
+		},
+	}
+	if err = ctrl.SetControllerReference(cluster, svc, r.Scheme); err != nil {
+		return fmt.Errorf("cannot set controller reference: %w", err)
+	}
+	if err = r.Create(ctx, svc); err != nil {
+		return fmt.Errorf("cannot create cluster state configmap: %w", err)
+	}
+	return nil
+}
+
+// ensureClusterStateConfigMap creates or updates cluster state configmap.
+func (r *EtcdClusterReconciler) ensureClusterStateConfigMap(
+	ctx context.Context, cluster *etcdaenixiov1alpha1.EtcdCluster) error {
+	configMap := &corev1.ConfigMap{}
+	err := r.Get(ctx, client.ObjectKey{
+		Namespace: cluster.Namespace,
+		Name:      r.getClusterStateConfigMapName(cluster),
+	}, configMap)
+	// configmap exists, skip editing.
+	if err == nil {
+		return nil
+	}
+
+	// configmap does not exist, create with cluster state "new"
+	if errors.IsNotFound(err) {
+		configMap = &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: cluster.Namespace,
+				Name:      r.getClusterStateConfigMapName(cluster),
+			},
+			Data: map[string]string{
+				"ETCD_INITIAL_CLUSTER_STATE": "new",
+			},
+		}
+		if err := ctrl.SetControllerReference(cluster, configMap, r.Scheme); err != nil {
+			return fmt.Errorf("cannot set controller reference: %w", err)
+		}
+		if err := r.Create(ctx, configMap); err != nil {
+			return fmt.Errorf("cannot create cluster state configmap: %w", err)
+		}
+		return nil
+	}
+	return fmt.Errorf("cannot get cluster state configmap: %w", err)
+}
+
+func (r *EtcdClusterReconciler) ensureClusterStatefulSet(
+	ctx context.Context, cluster *etcdaenixiov1alpha1.EtcdCluster) error {
+	statefulSet := &appsv1.StatefulSet{}
+	err := r.Get(ctx, client.ObjectKey{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Name,
+	}, statefulSet)
+
+	// statefulset does not exist, create new one
+	notFound := false
+	if errors.IsNotFound(err) {
+		notFound = true
+		// prepare initial cluster members
+		initialCluster := ""
+		for i := uint(0); i < cluster.Spec.Replicas; i++ {
+			if i > 0 {
+				initialCluster += ","
+			}
+			initialCluster += fmt.Sprintf("%s-%d=https://%s-%d.%s.%s.svc.cluster.local:2380",
```

Comment:
Please do not hardcode `cluster.local` suffixes as many clusters may have different ones.

It's possible to access services without them.